src/manager/evm.rs

```diff
@@ -26,127 +30,213 @@ pub type MiddlewareImpl = SignerMiddleware<Provider<Http>, Wallet<SigningKey>>;
+            min_activation_collateral: ethers::types::U256::from(min_validator_stake as u128),
+            min_validators: params.min_validators,
+            bottom_up_check_period: params.bottomup_check_period as u64,
+            top_down_check_period: params.topdown_check_period as u64,
+            // TODO: update this variable properly
+            majority_percentage: 50,
+            genesis: ethers::types::Bytes::default(),
+        };
+        log::info!("creating subnet: {params:?}");
+
+        let call = self.registry_contract.new_subnet_actor(params);
+        let pending_tx = call.send().await?;
+        let receipt = pending_tx.retries(10).await?;
+        return match receipt {
+            Some(r) => {
+                for log in r.logs {
+                    log::debug!("log: {log:?}");
+
+                    match ethers_contract::parse_log::<ipc_registry::SubnetDeployedFilter>(log) {
+                        Ok(subnet_deploy) => {
+                            let ipc_registry::SubnetDeployedFilter {
+                                subnet_addr,
+                                subnet_id,
+                            } = subnet_deploy;
+
+                            log::debug!("subnet with id {subnet_id:?} deployed at {subnet_addr:?}");
+
+                            // subnet_addr.to_string() returns a summary of the actual address, not
+                            // usable in the actual code.
+                            let subnet_addr = format!("{subnet_addr:?}");
+                            log::debug!("raw subnet addr: {subnet_addr:}");
+
+                            let eth_addr = EthAddress::from_str(&subnet_addr)?;
+                            return Ok(Address::from(eth_addr));
+                        }
+                        Err(_) => {
+                            log::debug!("not of event subnet actor deployed, continue");
+                            continue;
+                        }
+                    }
+                }
+                Err(anyhow!("no logs receipt"))
+            }
+            None => Err(anyhow!("no receipt to event, txn not successful")),
+        };
     }
 
     async fn join_subnet(
         &self,
         subnet: SubnetID,
-        from: Address,
+        _from: Address,
         collateral: TokenAmount,
-        params: JoinParams,
+        _params: JoinParams,
     ) -> Result<()> {
-        todo!()
+        let collateral = collateral
+            .atto()
+            .to_u64()
+            .ok_or_else(|| anyhow!("invalid min validator stake"))?;
+
+        let address = agent_subnet_to_evm_address(&subnet)?;
+        log::info!(
+            "interacting with evm subnet contract: {address:} with collateral: {collateral:}"
+        );
+
+        let contract = SubnetContract::new(address, self.eth_client.clone());
+
+        let mut txn = contract.join();
+        txn.tx.set_value(collateral);
+
+        txn.send().await?.await?;
```

Comment:
If the transaction doesn't go through this should fail, right?

Reply:
yep, it will fail.